both/access_control.js

```diff
@@ -231,101 +229,75 @@ if(Meteor.isServer) {
         'createdAt',
       ];
 
-      if(_.difference(fields, allowedFields).length > 0) {
+      return _onlyAllowedFields(fields, allowedFields);
+    },
+    fetch: [ 'competitionId' ],
+  });
+
+  RoundProgresses.allow({
+    update: function(userId, progress, fields, modifier) {
+      if(getCannotManageCompetitionReason(userId, progress.competitionId)) {
         return false;
       }
-      return true;
+      var allowedFields = [
+        'done',
+        'total',
+      ];
+      return _onlyAllowedFields(fields, allowedFields);
     },
-    fetch: [ 'competitionId' ],
   });
 
   Results.allow({
     update: function(userId, result, fields, modifier) {
-      var competition = Competitions.findOne({ _id: result.competitionId }, { fields: { _id: 1 } });
-      if(getCannotManageCompetitionReason(userId, competition._id)) {
+      if(getCannotManageCompetitionReason(userId, result.competitionId)) {
         return false;
       }
-
       var allowedFields = [
         'solves',
 
         'updatedAt',
         'createdAt',
       ];
-
-      if(_.difference(fields, allowedFields).length > 0) {
-        return false;
-      }
-      return true;
+      return _onlyAllowedFields(fields, allowedFields);
     },
     fetch: [ 'competitionId' ],
   });
 
-  var allowedOwnRegistrationFields = [
-    'userId',
-    'competitionId',
-    'uniqueName',
-    'registeredEvents',
-    'guestCount',
-    'comments',
-
-    'updatedAt',
-    'createdAt',
-  ];
   Registrations.allow({
     insert: function(userId, registration) {
-      if(!getCannotManageCompetitionReason(userId, registration.competitionId)) {
-        // If you're allowed to manage the competition, then you can
-        // add any registration.
-        return true;
-      }
-      if(getCannotRegisterReasons(registration.competitionId)) {
-        return false;
-      }
-      // can only edit entries with their user id
-      if(registration.userId != userId) {
-        return false;
-      }
-
-      if(_.difference(fields, allowedOwnRegistrationFields).length > 0) {
-        return false;
-      }
-      return true;
+      return Registrations.userIsAllowed(userId, registration);
```

Comment:
1. I didn't realize that this is the same as #209, whoops!
2. I think it's completely reasonable to check the fields on an document getting inserted by looking at `Object.keys(...)`. Updates are different, because you get the old document, and a list of fields the update is trying to change in that document.

I don't have this code in my mental cache right now, but agreed that we should not be listing the fields in more than one place. I don't necessarily agree that just because a client isn't allowed to write to something ("treated differently"), it means it should be in a different collection.

I'm probably preaching to the choir, but I think this code would get a lot simpler if we were using methods instead of access rules like this.